File: text-justification/csharp/src/TextJustification/TextJustifier.cs
using System.Text;

namespace TextJustification;

public static class TextJustifier
{
    public static IReadOnlyList<string> Justify(string text, int width)
    {
        var words = text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (words.Length == 0)
        {
            return Array.Empty<string>();
        }

        var lines = new List<string>();
        var lineWords = new List<string>();
        var lineContentLength = 0;

        foreach (var word in words)
        {
            var projected = lineContentLength + lineWords.Count + word.Length;
            if (lineWords.Count > 0 && projected > width)
            {
                lines.Add(JustifyLine(lineWords, lineContentLength, width));
                lineWords.Clear();
                lineContentLength = 0;
            }
            lineWords.Add(word);
            lineContentLength += word.Length;
        }

        if (lineWords.Count > 0)
        {
            lines.Add(string.Join(' ', lineWords));
        }

        return lines;
    }

    private static string JustifyLine(List<string> lineWords, int contentLength, int width)
    {
        if (lineWords.Count == 1)
        {
            var only = lineWords[0];
            return only.Length >= width ? only : only.PadRight(width);
        }

        var gaps = lineWords.Count - 1;
        var padding = width - contentLength;
        var baseSpaces = padding / gaps;
        var extras = padding % gaps;

        var sb = new StringBuilder(width);
        for (var i = 0; i < lineWords.Count; i++)
        {
            sb.Append(lineWords[i]);
            if (i < gaps)
            {
                var spaces = baseSpaces + (i < extras ? 1 : 0);
                sb.Append(' ', spaces);
            }
        }
        return sb.ToString();
    }
}

File: text-justification/typescript/src/textJustifier.ts
export function justify(text: string, width: number): readonly string[] {
  const words = text.split(/\s+/).filter((w) => w.length > 0);
  if (words.length === 0) {
    return [];
  }

  const lines: string[] = [];
  let lineWords: string[] = [];
  let lineContentLength = 0;

  for (const word of words) {
    const projected = lineContentLength + lineWords.length + word.length;
    if (lineWords.length > 0 && projected > width) {
      lines.push(justifyLine(lineWords, lineContentLength, width));
      lineWords = [];
      lineContentLength = 0;
    }
    lineWords.push(word);
    lineContentLength += word.length;
  }

  if (lineWords.length > 0) {
    lines.push(lineWords.join(' '));
  }

  return lines;
}

function justifyLine(lineWords: string[], contentLength: number, width: number): string {
  if (lineWords.length === 1) {
    const only = lineWords[0]!;
    return only.length >= width ? only : only + ' '.repeat(width - only.length);
  }

  const gaps = lineWords.length - 1;
  const padding = width - contentLength;
  const baseSpaces = Math.floor(padding / gaps);
  const extras = padding % gaps;

  let out = '';
  for (let i = 0; i < lineWords.length; i++) {
    out += lineWords[i];
    if (i < gaps) {
      const spaces = baseSpaces + (i < extras ? 1 : 0);
      out += ' '.repeat(spaces);
    }
  }
  return out;
}
